src/lib/logical_query_plan/multi_predicate_join_node.hpp

```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <memory>
+#include <optional>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "types.hpp"
+
+#include "abstract_lqp_node.hpp"
+#include "expression/abstract_expression.hpp"
+#include "lqp_column_reference.hpp"
+
+namespace opossum {
+
+class MultiPredicateJoinNode : public EnableMakeForLQPNode<MultiPredicateJoinNode>, public AbstractLQPNode {
```

Comment:
Sorry to burst in here, but just saw the commit in a mail:

IMO, the `JoinNode` should be extended with the ability to hold multiple predicates, instead of introducing a new JoinNode-type. The Join node's `join_predicate` parameter could just become a `join_predicates` parameter.

Or is there a specific reason why this wouldn't fit into the existing architecture? Also, predicates in the LQP should always be represented as `AbstractExpressions`.

Reply:
+1 for extending the join node.

But one question to you @mrzzzrm. Where do you see the right place to decide which of the predicates is the main join predicate and which ones are post-processed?

Reply:
The JoinOrderingRule has always decided the order of join predicates (including which are primary and which are secondary). See `AbstractJoinOrderingAlgorithm::_add_join_to_plan`.

Here, instead of creating one JoinNode and multiple PredicateNodes, just create a single JoinNode with the predicates in the preferred order (the first being the primary join predicate)

```c++
  // Build JoinNode (for primary predicate) and subsequent scans (for secondary predicates)
  auto plan_with_joins = std::shared_ptr<AbstractLQPNode>{};
  if (primary_join_predicate) {
    plan_with_joins = JoinNode::make(JoinMode::Inner, primary_join_predicate, left_lqp, right_lqp);
  } else {
    plan_with_joins = JoinNode::make(JoinMode::Cross, left_lqp, right_lqp);
  }

  for (const auto& predicate_and_cost : join_predicates_and_cost) {
    plan_with_joins = PredicateNode::make(predicate_and_cost.first, plan_with_joins);
  }

  return plan_with_joins;
```

Reply:
@Felerius , @janetzki , @pascalhpi : do you see any problems with running before the JoinReorderingRule?